gpt4all-chat/tests/python/test_server_api.py

```diff
@@ -83,5 +116,132 @@ def chat_server(chat_server_config: dict[str, str]) -> Iterator[None]:
+        response = request.post('models/foo', data=None, wait=True)
+    assert excinfo.value.response.status_code == 405
+
+    # GET for completions
+    with pytest.raises(HTTPError) as excinfo:
+        response = request.get('completions', wait=True)
+    assert excinfo.value.response.status_code == 405
+
+    # GET for chat completions
+    with pytest.raises(HTTPError) as excinfo:
+        response = request.get('chat/completions', wait=True)
+    assert excinfo.value.response.status_code == 405
+
+
+EXPECTED_MODEL_INFO = {
+    'created': 0,
+    'id': 'Llama 3.2 1B Instruct',
+    'object': 'model',
+    'owned_by': 'humanity',
+    'parent': None,
+    'permissions': [
+        {
+            'allow_create_engine': False,
+            'allow_fine_tuning': False,
+            'allow_logprobs': False,
+            'allow_sampling': False,
+            'allow_search_indices': False,
+            'allow_view': True,
+            'created': 0,
+            'group': None,
+            'id': 'placeholder',
+            'is_blocking': False,
+            'object': 'model_permission',
+            'organization': '*',
+        },
+    ],
+    'root': 'Llama 3.2 1B Instruct',
+}
+
+EXPECTED_COMPLETIONS_RESPONSE = {
+    'choices': [
+        {
+            'finish_reason': 'stop',
+            'index': 0,
+            'logprobs': None,
+            'references': None,
+            'text': ' jumps over the lazy dog.',
+        },
+    ],
+    'id': 'placeholder',
+    'model': 'Llama 3.2 1B Instruct',
+    'object': 'text_completion',
+    'usage': {
+        'completion_tokens': 6,
+        'prompt_tokens': 5,
+        'total_tokens': 11,
+    },
+}
+
+
+def test_with_models(chat_server_with_model: None) -> None:
+    response = request.get('models', wait=True)
+    assert response == {
+        'data': [EXPECTED_MODEL_INFO],
+        'object': 'list',
+    }
+
+    # Test the specific model endpoint
+    response = request.get('models/Llama 3.2 1B Instruct', wait=True)
+    assert response == EXPECTED_MODEL_INFO
+
+    # Test the completions endpoint
+    with pytest.raises(HTTPError) as excinfo:
+        request.post('completions', data=None, wait=True)
+    assert excinfo.value.response.status_code == 400
+
+    data = {
+        'model': 'Llama 3.2 1B Instruct',
+        'prompt': 'The quick brown fox',
+        'temperature': 0,
+    }
+
+    response = request.post('completions', data=data, wait=True)
+    assert 'choices' in response
+    assert response['choices'][0]['text'] == ' jumps over the lazy dog.'
+    assert 'created' in response
+    response.pop('created')  # Remove the dynamic field for comparison
+    assert response == EXPECTED_COMPLETIONS_RESPONSE
+
+    data['temperature'] = 0.5
+
+    pytest.xfail('This causes an assertion failure in the app. See https://github.com/nomic-ai/gpt4all/issues/3133')
+    with pytest.raises(HTTPError) as excinfo:
+        response = request.post('completions', data=data, wait=True)
```

Comment:
I believe this asserts that the *entire* test fails, not just this part. Checks that are expected to fail should be split into their own test functions.

Reply:
Yeah and typically you wouldn't invoke pytest.xfail like this -- instead, you'd create a small new test like this.

```
@pytest.mark.xfail
def test_temperature_setting(requestor):
    data = {
        'model': 'Llama 3.2 1B Instruct',
        'prompt': 'The quick brown fox',
        'temperature': 0.5,
    }
    with pytest.raises(HTTPError) as excinfo:
        response = request.post('completions', data=data, wait=True)
```